### DISCUSSION TITLE
Better controller support on linux is needed

### DISCUSSION THREAD

**Discussion by SandWoodJones** [ikemen-engine/Ikemen-GO#861] [Feature Request] [answered]
I'm using Arch Linux, and been having problems with setting up Ikemen-Go to use my DS4 controller.
First of all, I can't change the controller configuration through the in-game joystick config GUI as it doesn't capture any input from my controller, making me unable to edit it.
To fix this, I had to use `jstest` to get the ID number of each button and then manually map it to the config.json file. This also had problems since I couldn't figure out the ID numbers for the joystick and the D-Pad as they're axis, not buttons. Strangely enough the only movement I could get was moving left, by pressing Up on the D-Pad.
Then I decided to just add the game to my steam library as a non-steam game and use its built-in controller remapping, mapping each button and axis to a keyboard key.
This works, but it's not perfect nor optimal.

**Comment by ssj17vegeta** [comment]
Can confirm the current version is somewhat broken in terms of controller support (Ubuntu 21.04 here).

Plugging-in a Madcatz XBox controller crashes the program on startup and outputs this in the console :
`GLFW: An uncaught error has occurred: InvalidValue: Invalid button in gamepad mapping 050000005e040000fd02000003090000 (Xbox One Wireless Controller)
GLFW: Please report this bug in the Go package immediately.`

Though it seems that error is being investigated here : https://github.com/go-gl/glfw/issues/292

Plugging in a standard wired XBox 360 controller outputs no error, but the controls are completely non-responsive : D-Pad up and down are mixed up in the main menu, and I cannot input any buttons when configuring the controller in the options menu.

Will definitely try the steam mapping tomorrow as a workaround though :)

**Comment by Orochikyocr** [comment]
Can confirm this. 
Qanba Cristal Clear and Generic PS3 sticks which both works on windows doesn't not work on Linux in the config input menu.
Also DS4, PS3 and my Bluetooth Marvo Gt62 controller doesn't work in the config menu.
All of them work if I configure them directly from the config.json.

**Comment by ssj17vegeta** [comment]
It seems the problem I was talking about ( https://github.com/glfw/glfw/issues/1763 ) has been fixed 15 days ago, and will be integrated in 3.3.5. There's hope :)

Also tried to fiddle with config.json to no avail : buttons are registered OK, but the it seems the D-Pad axes are considered buttons instead of axes so... yah :(

**Comment by SeongGino** [comment]
Huh, so I'm not the only one who noticed.

Updating to 0.98 made it so Ikemen stops crashing with a gamepad connected. But I tried to reconfigure two different Xinput controllers; neither register in the config menu, despite working enough to navigate to said menu.

**Comment by Orochikyocr** [comment]
Yes, 0.98, my sticks are still broken. Ps3 and xbox 360 one. My generic Marvo Controller works out of the box.
The same situation, I can configure my sticks from the config.json and they work, except in the joystick input config menu.

**Comment by ssvb** [comment]
The [joysticks](https://github.com/glfw/glfw/blob/master/tests/joysticks.c) test program from https://github.com/glfw/glfw shows the following for my PS3 controller in Linux in an idle state (no buttons are pressed):
![ps3_idle_state](https://user-images.githubusercontent.com/78642/183241854-f84cc236-26da-4e5d-9168-7d6331442a27.png)

The sliders in the leftmost state (highlighted by the arrows) correspond to L2/R2 analog buttons of the PS3 controller. If I slowly press these buttons, then the sliders gradually move to the right and reach the rightmost side when the buttons are fully pressed.

Now how is all of this related to the broken joystick config in Ikemen-GO? The code at https://github.com/ikemen-engine/Ikemen-GO/blob/0.98.2/src/script.go#L1291-L1335 checks the joystick button state when configuring the joystick. In an idle PS3 controller state with no buttons pressed, the array `axes` typically looks like `[-0.0039215684 0.027451038 -1 0.011764765 -0.035294116 -1]`. The elements with indexes 2 and 5 contain `-1`, which is interpreted by the code as if some of the buttons are always pressed (the return value is "-11"). And then this return value is used by code at https://github.com/ikemen-engine/Ikemen-GO/blob/0.98.2/external/script/options.lua#L1754-L1761, which expects to see that some joystick button gets pressed and then all buttons are released. If it can never see that all buttons are released, then it can't work properly.

I have a pull request with a fix for my PS3 controller: https://github.com/ikemen-engine/Ikemen-GO/pull/629. The other controllers probably fail because of something similar. But I wonder why people are not complaining about Windows? Also the code seems to already have some special handling for xbox 360 controller and yet people reported problems with it in Linux.

**Comment by leonkasovan** [comment]
i also notice this.
so i made this fix in file: input.go

```go
func JoystickState(joy, button int) bool {
	if joy < 0 {
		return sys.keyState[Key(button)]
	}
	if joy >= input.GetMaxJoystickCount() {
		return false
	}
	var axes []float32
	if button >= 0 {
		// Query button state
		btns := input.GetJoystickButtons(joy)
		if button >= len(btns) {
			return false
		}
		switch button {
		case 10:	// Up: check axis and d.pad(hat)
			axes = input.GetJoystickAxes(joy)
			return (axes[1] < -0.5)|| (btns[11] != 0)
		case 11:	// Right: check axis and d.pad(hat)
			axes = input.GetJoystickAxes(joy)
			return (axes[0] > 0.5) || (btns[12] != 0)
		case 12:	// Down: check axis and d.pad(hat)
			axes = input.GetJoystickAxes(joy)
			return (axes[1] > 0.5) || (btns[13] != 0)
		case 13:	// Left: check axis and d.pad(hat)
			axes = input.GetJoystickAxes(joy)
			return (axes[0] < -0.5) || (btns[14] != 0)
		default:	// Other (normal) button
			return btns[button] != 0
		}
	} else {
		axes = input.GetJoystickAxes(joy)
		if len(axes) < 6 {
			return false
		}
		switch button {
		case -12:
			return (axes[2] > -0.5)
		case -10:
			return (axes[5] > -0.5)
		default:
			return false
		}
	}
}
```

That code fix:
1. D-Pad up and down are mixed up in the main menu
2. Left thumb stick 
3. L2 and R2 

i tested Native Linux Build in Steam Deck using SteamOS.
The linux binary can be downloaded from here https://github.com/leonkasovan/Ikemen-GO/releases/tag/v.0.99.0-steamdeck

**Reply by ssvb** [reply]
@leonkasovan You don't need to patch the code to remap the joystick buttons. Editing `save/config.json` should be enough (or doing the joystick configuration inside of the game via the options menu). For my PS3 controller, I'm using the following configuration:
```json
  "JoystickConfig": [
    {
      "Joystick": 0,
      "Buttons": [
        "13",
        "14",
        "15",
        "16",
        "0",
        "1",
        "5",
        "3",
        "2",
        "4",
        "9",
        "7",
        "6",
        "8"
      ]
    },
```

The https://github.com/ikemen-engine/Ikemen-GO/pull/629 pull request had been already merged. It fixed the game "freeze" in the joystick configuration menu in the game.

Currently Ikemen-Go doesn't have a sane joystick buttons configuration out of the box and this is definitely not user friendly. But it's at least configurable by the user. Yes, it would make sense to additionally have some initial joystick configuration presets (based on the Hardware GUIDs). See https://github.com/ikemen-engine/Ikemen-GO/discussions/920

**Reply by ssvb** [reply]
If your symptom is *"can't change the controller configuration through the in-game joystick config GUI as it doesn't capture any input from my controller"* as mentioned in the starter post of this thread and your joystick is not PS3, then you probably need a fix similar to https://github.com/ikemen-engine/Ikemen-GO/pull/629 (which resolved the problem for PS3).

But if you just have some buttons swapped, then it's just a configuration issue.

**Reply by leonkasovan** [reply]
by editing `save/config.json` can solve "left thumb stick" too? mine does not work at all. maybe there is another workaround?

**Reply by ssvb** [reply]
What's your joystick hardware? It's possible to configure either "d-pad" or the "left thumb stick" for input and it should work. Or do you expect that both "d-pad" and "left thumb stick" should work at the same time and duplicate each other's functionality?

**Comment by leonkasovan** [comment]
> What's your joystick hardware? It's possible to configure either "d-pad" or the "left thumb stick" for input and it should work. Or do you expect that both "d-pad" and "left thumb stick" should work at the same time and duplicate each other's functionality?

I use Ikemen_GO_Linux (v0.99.0) in Steamdeck (SteamOS).
I expect d-pad and left thumb stick work at same time.
How do we configure the setting in save/config.json?

**Comment by ssvb** [comment]
> I expect d-pad and left thumb stick work at same time. How do we configure the setting in save/config.json?

You can't configure this in config.json right now. So patching the code would be necessary to make this possible. However your patch is tailored for your specific joystick hardware and ruins joystick support for everyone else, so it's unlikely to be accepted. The magic button id numbers are even different in Windows and Linux.

**Comment by jmfergeau** [comment]
same problem. can't configure the gamepad. It's recognized by the game but the directions are all wrong with left and right mapped to up and down. (!)

it's a generic xbox one gamepad (that works everywhere else) using the xone dkms driver. I'm on arch linux.

I tried to check the names of the buttons with jstest but i have no idea what i'm supposed to put for the DPad. Axis6+? 

Only workaround i got is to disable completely the joystick (by putting `Not used` on all buttons) in the config file then use antimicro to use the keyboard as a joypad

**Reply by ssvb** [reply]
@jmfergeau What do you mean by "can't configure"? What exactly are you doing? You should be able to use the keyboard to navigate the main menu and go to "OPTIONS" -> "Input Settings" -> "Joystick Config" -> "Config all". Press the buttons of your joystick (DPad for the Up/Down/Left/Right) until everything is configured. Then don't forget to "Save and Return" in "OPTIONS".

The joystick buttons are not configured properly out of the box, because there are so many different joystick models and different drivers in different operating systems interpret them in different ways. Yes, this makes the game look somewhat unpolished and gives a bad impression to the players, who don't like tinkering and just want to immediately enjoy the game. Not sure if @K4thos has anything to say about this and what are the current priorities.

**Comment by jmfergeau** [comment]
> @jmfergeau What do you mean by "can't configure"? What exactly are you doing?

Exactly like i do with mugen for around 20 years now, which is like you described, OPTIONS" -> "Input Settings" -> "Joystick Config" -> "Config all". Except nothing happens when i press any buttons or sticks when i do that. Note that the joystick replies for the rest, like navigating menus and actually play. Though it's really badly mapped.

**Comment by leonkasovan** [comment]
> same problem. can't configure the gamepad. It's recognized by the game but the directions are all wrong with left and right mapped to up and down. (!)
> 
> it's a generic xbox one gamepad (that works everywhere else) using the xone dkms driver. I'm on arch linux.
> 
> I tried to check the names of the buttons with jstest but i have no idea what i'm supposed to put for the DPad. Axis6+?
> 
> Only workaround i got is to disable completely the joystick (by putting `Not used` on all buttons) in the config file then use antimicro to use the keyboard as a joypad

what's system are you using for Ikemen/Mugen? PC? Windows? Linux? Handheld? Console?
How to configure antimicro for Ikemen/Mugen?

**Reply by jmfergeau** [reply]
Arch Linux with LTS kernel.

Here, I joined the config I made. It's not perfect, depending of the characters, but it's better than nothing. (It's for antimicrox. I'm not sure it's compatible with antimicro but i don't see why it wouldn't be)
[ikemengo.gamecontroller.zip](https://github.com/ikemen-engine/Ikemen-GO/files/14376446/ikemengo.gamecontroller.zip)

**Comment by Jesuszilla** [comment]
Continuing this discussion because mapping gamepads on RetroPie (Raspbian, Raspberry Pi 5, ARM64) is really inconsistent. It detected my RetroBit Genesis 2.4 GHz pad temporarily last night and now I can't seem to get IKEMEN-GO to detect it again, let alone remap it. What's odd is, I can hit an oddly specific button to "Configure all" (B) on my pad, but pressing that same button again isn't picked up by the joystick mapping configuration to assign a button! It won't even detect my directions for some reason!

EDIT: The issue seems to be that it thinks my gamepad should be assigned to Player 3 for some reason. ~~Why can't I assign it to Player 2 like I want? This is a major accessibility issue.~~

EDIT2: Just set your Joystick index in your config.json to the index of the working joystick for the player you want to configure it for.
We should really have an in-engine way of selecting it though imo

**Comment by leonkasovan** [comment]
> Continuing this discussion because mapping gamepads on RetroPie (Raspbian, Raspberry Pi 5, ARM64) is really inconsistent. It detected my RetroBit Genesis 2.4 GHz pad temporarily last night and now I can't seem to get IKEMEN-GO to detect it again, let alone remap it. What's odd is, I can hit an oddly specific button to "Configure all" (B) on my pad, but pressing that same button again isn't picked up by the joystick mapping configuration to assign a button! It won't even detect my directions for some reason!
> 
> EDIT: The issue seems to be that it thinks my gamepad should be assigned to Player 3 for some reason. ~Why can't I assign it to Player 2 like I want? This is a major accessibility issue.~
> 
> EDIT2: Just set your Joystick index in your config.json to the index of the working joystick for the player you want to configure it for. We should really have an in-engine way of selecting it though imo

Could you show me the setting in config.json that work for you?

Mine like this and does not work on SteamOS :
```
        "JoystickConfig": [
                {
                        "Joystick": 0,
                        "Buttons": [
                                "10",
                                "11",
                                "13",
                                "12",
                                "1",
                                "0",
                                "5",
                                "3",
                                "2",
                                "4",
                                "-10",
                                "Not used",
                                "Not used",
                                "7"
                        ]
                },
```

**Reply by Jesuszilla** [reply]
Directly to the right of "Joystick" in your example is a number. You may need to set it to 1,2, or even 3. Try different indices.

**Comment by K4thos** [comment, marked as answer]
I think this feature request is no longer relevant, we do have better controllers system using sdl db
